### PR TITLE
Expose populateReshapeToInterfaceTensorPatterns and use in rank-reduc…

### PR DESCRIPTION
**Pull request by nicolasvasilache**
…ing pattern application

This is useful to better cleanup rank-reducing IR.

### CODE REVIEW

**Review by ThomasRaoux — APPROVED** [review]

**Review by MaheshRavishankar — COMMENTED** [review]
I wonder why the `BUILD` file is changed this way. I see `CommonPasses` is used in the `TransformExtensions`) but I dont understand the demarcation between what is in `Common` and what is in `CommonPasses`. Also `TileAndDistributeToWorkgroups` depends on `TileDispatchUsingInterface`. The latter is removed from `CommonPasses` but is added to `Common`. The former is there in both. I am a bit lost regarding the `BUILD` file changes.

**Comment by nicolasvasilache** [comment]
The TL;DR is that Utils, canonicalization patterns/populate, transform patterns/populate and passes were all crumpled together in a single library.

Extracting a canonicalization patterns/populate for reuse immediately turned into cyclic dependencies and I didn't want to just copy code around.

I am happy to iterate to a more modular breakdown along the following axes (each becoming a separate library): 1) utils, 2) canonicalization patterns/populate, 3) transform patterns/populate and 4) passes.

Then Common could depend on those 4 things.

This will be quite more intrusive so for now I split along rough lines between things that "looks like passes" and the rest.
Since the C++ code is still crumpled, these rough boundaries may not make as much sense but have the merit of breaking the cyclic dependence.

Does the above 1-4) breakdown sound reasonable?
